proposals/middleware-in-integrations.md

```diff
@@ -0,0 +1,128 @@
+          order: 'pre'
+        });
+      }
+    }
+  };
+}
+```
+
+This API takes from `addRenderer`, `addClientDirective`, and `setAdapter`, to provide an object with an entrypoint. That entrypoint will be loaded by Vite and wired up with the user's own middleware.
+
+# Background & Motivation
+
+When we initially designed middleware we took a integration-first approach, but quickly realized that this made it difficult for a user to add their own middleware. So we decided to punt on integration support and nail down the user API first.
+
+There are use-cases where a package might want to provide both an integration *and* middleware. Currently the user has to do this by adding the middleware themselves, and then also add an integration.
+
+We've had requests for this feature for [Starlight](https://starlight.astro.build/), as well as for auth libraries.
+
+# Goals
+
+- Allow integrations to transparently add middleware.
+- Allow the middleware to be placed either at the front, where it would run *first*, or the back, where it would run *last*.
+
+# Non-Goals
+
+- Future: the ability for integrations to fully control the order. Vite plugins, which are similar, allow this through a `configResolved` hook. There might be some use-cases for this with middleware in the future.
+
+# Detailed Design
+
+The API of this works very similar to how `addRenderer`, `addClientDirective`, and `setAdapter` works. Both of these APIs are called by an integration and then the information about them is stored on the internal `settings` object.
+
+In this case there will be an object that looks like `middleware: { pre: [], post: [] }` on settings. The `addMiddleware` implementation will push the entrypoint for each integration call into the appropriate stack.
+
+This means that the order, when multiple integrations use the same `order`, will be the order in which integrations are added:
+
+```js
+export default defineConfig({
+  integrations: [
+    one(),
+    two()
+  ]
+})
+```
+
+In the above case, if both integrations add a `pre` middleware, the underlying data structure will look like:
+
+```js
+{
+  middleware: {
+    pre: [one, two],
+    post: []
+  }
+}
+```
+
+This means that users have ultimate control over order by deciding which order integrations should be added.
+
+In reality conflicts should be rare as middleware authors try to be good citizens and only affect requests that are specific to their needs.
+
+## Runtime implementation
+
+Once the `settings` object contains the ordered list of middleware, a Vite plugin will be used to create a single middleware `onRequest` function using the `sequence` API. The pseudo-code of what this plugin produces looks like so:
+
+```js
+import { sequence } from 'astro:middleware';
+import { onRequest as userOnRequest } from '/src/middleware';
+
+import { onRequest as _pre_0 } from 'one';
+import { onRequest as _pre_1 } from 'two';
+
+import { onRequest as _post_0 } from 'one';
```

Comment:
The same middleware runs twice per request. When might this be necessary?

Reply:
Is "both" a valid value for "order"?

Reply:
Seems like the example might be unintentional.

Reply:
No, both is not a valid value. This is just showing that an integration adds both a pre and a post. You have to add one or the other. I can make this be a different integration to make that clearer.

Reply:
Do you think the `astro:config:setup` hook example might be a better place to show `addMiddleware` being called multiple times?

Reply:
This example is implementation details. The idea is to lay out how you think the feature should be implemented so that someone could take the information and implement it. So it needs to be hear to convey the expected compile output.